// @pie-lib/text-select/lib/correctness-token.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _green = require('@material-ui/core/colors/green');

var _green2 = _interopRequireDefault(_green);

var _orange = require('@material-ui/core/colors/orange');

var _orange2 = _interopRequireDefault(_orange);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CorrectnessToken = function CorrectnessToken(props) {
  var classes = props.classes;


  var className = (0, _classnames2.default)(classes.custom, props.correct === true && classes.correct, props.correct === false && classes.incorrect);

  return _react2.default.createElement(
    'span',
    { className: className },
    props.text
  );
};

CorrectnessToken.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  correct: _propTypes2.default.bool,
  text: _propTypes2.default.string.isRequired
};

exports.default = (0, _styles.withStyles)(function () {
  return {
    correct: {
      backgroundColor: _green2.default[500]
    },
    incorrect: {
      backgroundColor: _orange2.default[500]
    }
  };
})(CorrectnessToken);
