// @pie-element/match-configure/src/configure.jsx
import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import { FeedbackConfig } from '@pie-lib/config-ui';
import PartialScoringConfig from '@pie-lib/scoring-config';
import SwipeableViews from 'react-swipeable-views';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import PropTypes from 'prop-types';
import debug from 'debug';
import Typography from '@material-ui/core/Typography';
import GeneralConfigBlock from './general-config-block';
import AnswerConfigBlock from './answer-config-block';

const log = debug('@pie-element:match:configure');

const styles = theme => ({
  title: {
    fontSize: '1.1rem',
    display: 'block',
    marginTop: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit
  },
  content: {
    marginTop: theme.spacing.unit * 2
  }
});

class Configure extends React.Component {
  static propTypes = {
    onModelChanged: PropTypes.func,
    classes: PropTypes.object,
    model: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);

    this.rowIdCounter = props.model.rows[props.model.rows.length - 1].id + 1;

    this.state = {
      activeTab: 0
    };
  }

  onTabChange = (event, value) => {
    this.setState({ activeTab: value });
  };

  onChangeTabIndex = index => {
    this.setState({ activeTab: index });
  };

  onChange = model => {
    this.props.onModelChanged(model);
  };

  onFeedbackChange = feedback => {
    const { model, onModelChanged } = this.props;
    model.feedback = feedback;
    onModelChanged(model);
  };

  onDeleteRow = (rowIndex) => {
    const { model } = this.props;

    const newModel = { ...model };

    newModel.rows.splice(rowIndex, 1);

    this.onChange(newModel);
  };

  onAddRow = () => {
    const { model } = this.props;
    const newModel = { ...model };

    newModel.rows = newModel.rows.concat({
      id: this.rowIdCounter + 1,
      title: `Question Text ${newModel.rows.length + 1}`,
      values: new Array(model.layout - 1).fill(false)
    });

    this.rowIdCounter += 1;

    this.onChange(newModel);
  };

  onLayoutChange = (newLayout) => {
    const { model } = this.props;
    const oldLayout = model.layout;
    const newModel = { ...model };

    if (newLayout > oldLayout) {
      for (let i = 0; i < (newLayout - oldLayout); i++) {
        newModel.headers.push(`Column ${newModel.headers.length + 1}`);
      }

      newModel.rows.forEach(row => {
        for (let i = 0; i < (newLayout - oldLayout); i++) {
          row.values.push(false);
        }
      });
    } else if (newLayout < oldLayout) {
      newModel.headers.splice(newLayout);

      newModel.rows.forEach(row => {
        row.values.splice(newLayout - 1);
      });

    }

    newModel.layout = newLayout;

    this.onChange(newModel);
  };

  onResponseTypeChange = (newResponseType) => {
    const { model } = this.props;
    const newModel = { ...model };

    // if we're switching to radio and we have more than one true, reset
    if (newResponseType === 'radio') {
      newModel.rows.forEach(row => {
        const trueCount = row.values.reduce((total, current) => current === true ? total + 1 : total);

        if (trueCount > 1) {
          row.values = new Array(model.layout - 1).fill(false)
        }
      })
    }

    newModel.responseType = newResponseType;

    this.onChange(newModel);
  };

  onPartialScoringChange = partialScoring => {
    this.props.model.partialScoring = partialScoring.map(partialScore => ({
      numberOfCorrect: partialScore.numberOfCorrect || '',
      scorePercentage: partialScore.scorePercentage || ''
    }));

    this.props.onModelChanged(this.props.model);
  };

  render() {
    const { classes, model } = this.props;

    log('[render] model', model);

    return (
      <div>
        <Tabs
          value={this.state.activeTab}
          onChange={this.onTabChange}
          indicatorColor="primary"
          textColor="primary"
          fullWidth
        >
          <Tab label="Design" />
          <Tab disabled={!model.allowPartialScoring} label="Scoring" />
        </Tabs>
        <SwipeableViews
          axis="x"
          index={this.state.activeTab}
          onChangeIndex={this.onChangeTabIndex}
        >
          <div className={classes.tab}>
            <div className={classes.content}>
              <Typography component="div" type="body1">
                <span>
                  In Choice Matrix, students associate choices in the first column with options in the adjacent rows.
                  This interaction allows for either one or more correct answers. Setting more than one answer as correct allows for partial credit <i>(see the Scoring tab)</i>.
                </span>
              </Typography>
              <GeneralConfigBlock
                model={model}
                onResponseTypeChange={this.onResponseTypeChange}
                onLayoutChange={this.onLayoutChange}
              />
              <AnswerConfigBlock
                model={model}
                onChange={this.onChange}
                onAddRow={this.onAddRow}
                onDeleteRow={this.onDeleteRow}
              />
              <FeedbackConfig
                feedback={model.feedback}
                onChange={this.onFeedbackChange}
              />
            </div>
          </div>
          <div className={classes.tab}>
            <PartialScoringConfig
              numberOfCorrectResponses={model.rows.length}
              partialScoring={model.partialScoring}
              onChange={this.onPartialScoringChange}
              />
          </div>
        </SwipeableViews>
      </div>
    );
  }
}

const ConfigureMain = withStyles(styles)(Configure);

class StateWrapper extends React.Component {
  static propTypes = {
    model: PropTypes.any,
    onModelChanged: PropTypes.func
  };

  constructor(props) {
    super(props);

    this.state = {
      model: props.model
    };

    this.onModelChanged = m => {
      this.setState({ model: m }, () => {
        this.props.onModelChanged(this.state.model);
      });
    };
  }

  render() {
    const { model } = this.state;
    return <ConfigureMain model={model} onModelChanged={this.onModelChanged} />;
  }
}

export default StateWrapper;
