// @pie-lib/math-input/lib/number-pad.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NumberPad = exports.defaults = exports.numbers = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _IconButton = require('@material-ui/core/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styles = require('./styles');

var _merge = require('lodash/merge');

var _merge2 = _interopRequireDefault(_merge);

var _styles2 = require('@material-ui/core/styles');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  root: {
    display: 'grid',
    gridTemplateColumns: 'repeat(3, 1fr)',
    gridRowGap: '0px',
    gridColumnGap: '0px'
  }
};

var numbers = exports.numbers = ['7', '8', '9', '4', '5', '6', '1', '2', '3', '0'];
var extras = ['.', '='];
var defaults = exports.defaults = numbers.concat(extras);

var baseStyles = (0, _merge2.default)((0, _styles.buttonStyle)(), {
  root: {
    backgroundColor: '#cacaca',
    height: '100%'
  }
});

var NumberPadButton = (0, _styles2.withStyles)(baseStyles)(function (props) {
  return _react2.default.createElement(
    _IconButton2.default,
    {
      tabIndex: '-1',
      onClick: function onClick() {
        props.onClick(props.value);
      },
      classes: props.classes
    },
    props.children
  );
});

var EqualsButton = (0, _styles2.withStyles)((0, _merge2.default)((0, _styles.buttonStyle)(), {
  root: {
    backgroundColor: 'orange',
    height: '100%'
  }
}))(function (props) {
  return _react2.default.createElement(NumberPadButton, props);
});

var NumberPad = exports.NumberPad = function (_React$Component) {
  _inherits(NumberPad, _React$Component);

  function NumberPad(props) {
    _classCallCheck(this, NumberPad);

    var _this = _possibleConstructorReturn(this, (NumberPad.__proto__ || Object.getPrototypeOf(NumberPad)).call(this, props));

    _this.onClick = _this.onClick.bind(_this);
    return _this;
  }

  _createClass(NumberPad, [{
    key: 'onClick',
    value: function onClick(value) {
      this.props.onClick(value);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          classes = _props.classes,
          values = _props.values;

      return _react2.default.createElement(
        'div',
        { className: classes.root },
        values.map(function (v) {
          var Button = v === '=' ? EqualsButton : NumberPadButton;
          return _react2.default.createElement(
            Button,
            { key: v, onClick: _this2.onClick, value: v },
            v
          );
        })
      );
    }
  }]);

  return NumberPad;
}(_react2.default.Component);

NumberPad.propTypes = {
  onClick: _propTypes2.default.func,
  classes: _propTypes2.default.object.isRequired,
  values: _propTypes2.default.arrayOf(_propTypes2.default.string)
};

NumberPad.defaultProps = {
  values: defaults
};

exports.default = (0, _styles2.withStyles)(styles)(NumberPad);
