// @pie-element/placement-ordering/configure/lib/main.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Tabs = require('@material-ui/core/Tabs');

var _Tabs2 = _interopRequireDefault(_Tabs);

var _Tab = require('@material-ui/core/Tab');

var _Tab2 = _interopRequireDefault(_Tab);

var _design = require('./design');

var _design2 = _interopRequireDefault(_design);

var _help = require('./help');

var _help2 = _interopRequireDefault(_help);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _scoringConfig = require('@pie-lib/scoring-config');

var _scoringConfig2 = _interopRequireDefault(_scoringConfig);

var _placementOrdering = require('@pie-ui/placement-ordering');

var _styles = require('@material-ui/core/styles');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _cloneDeep = require('lodash/cloneDeep');

var _cloneDeep2 = _interopRequireDefault(_cloneDeep);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = {
  tabBar: {
    display: 'flex',
    justifyContent: 'space-between'
  }
};

class Main extends _react2.default.Component {

  constructor(props, context) {
    super(props, context);

    this.state = {
      index: 0,
      model: props.initialModel
    };

    this.onTabIndexChange = (event, index) => {
      this.setState({ index });
    };

    this.onPartialScoringChange = partialScoring => {
      const onModelChange = this.props.onModelChange;

      const model = (0, _cloneDeep2.default)(this.state.model);
      model.partialScoring = partialScoring;
      this.setState({ model }, () => {
        onModelChange(this.state.model);
      });
    };

    this.onModelChange = model => {
      const onModelChange = this.props.onModelChange;

      const resetSession = model.placementType !== this.state.model.placementType || model.choices.length !== this.state.model.choices.length;
      this.setState({ model }, () => {
        onModelChange(this.state.model, resetSession);
      });
    };
  }

  render() {
    var _props = this.props;
    const classes = _props.classes,
          imageSupport = _props.imageSupport;
    var _state = this.state;
    const index = _state.index,
          model = _state.model;


    return _react2.default.createElement(
      'div',
      null,
      _react2.default.createElement(
        'div',
        { className: classes.tabBar },
        _react2.default.createElement(
          _Tabs2.default,
          { onChange: this.onTabIndexChange, value: index },
          _react2.default.createElement(_Tab2.default, { label: 'Design' }),
          _react2.default.createElement(_Tab2.default, { label: 'Scoring' })
        ),
        _react2.default.createElement(_help2.default, null)
      ),
      index === 0 && _react2.default.createElement(_design2.default, {
        model: model,
        onModelChange: this.onModelChange,
        imageSupport: imageSupport
      }),
      index === 1 && _react2.default.createElement(_scoringConfig2.default, {
        partialScoring: model.partialScoring,
        numberOfCorrectResponses: model.correctResponse.length,
        onChange: this.onPartialScoringChange
      })
    );
  }
}

Main.propTypes = {
  initialModel: _propTypes2.default.object,
  onModelChange: _propTypes2.default.func.isRequired,
  classes: _propTypes2.default.object.isRequired,
  imageSupport: _propTypes2.default.object
};
exports.default = (0, _placementOrdering.withDragContext)((0, _styles.withStyles)(styles)(Main));
