// @pie-ui/categorize/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _mathRendering = require('@pie-lib/math-rendering');

var _piePlayerEvents = require('@pie-framework/pie-player-events');

var _categorize = require('./categorize');

var _categorize2 = _interopRequireDefault(_categorize);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Categorize extends HTMLElement {
  set model(m) {
    this._model = m;

    this.dispatchEvent(new _piePlayerEvents.ModelSetEvent(this.tagName.toLowerCase(), this.isComplete(), !!this._model));
    this.render();
  }

  isComplete() {
    if (!this._session) {
      return false;
    }

    return Array.isArray(this._session.answers) && this._session.answers.length > 0;
  }

  set session(s) {
    if (s && !s.answers) {
      s.answers = [];
    }

    this._session = s;
    this.render();
  }

  changeAnswers(answers) {
    this._session.answers = answers;

    this.dispatchEvent(new _piePlayerEvents.SessionChangedEvent(this.tagName.toLowerCase(), this.isComplete()));

    this.render();
  }

  render() {
    if (this._model && this._session) {
      const el = _react2.default.createElement(_categorize2.default, {
        model: this._model,
        session: this._session,
        onAnswersChange: this.changeAnswers.bind(this)
      });
      _reactDom2.default.render(el, this, () => {
        (0, _mathRendering.renderMath)(this);
      });
    }
  }
}
exports.default = Categorize;
