// @pie-ui/graph-lines/lib/controls.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Controls = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Button = require('@material-ui/core/Button');

var _Button2 = _interopRequireDefault(_Button);

var _styles = require('@material-ui/core/styles');

var _Delete = require('@material-ui/icons/Delete');

var _Delete2 = _interopRequireDefault(_Delete);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Controls extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          disabled = _props.disabled,
          onDeleteClick = _props.onDeleteClick,
          iconOnly = _props.iconOnly;


    return _react2.default.createElement(
      _Button2.default,
      {
        className: iconOnly ? classes.smallButton : classes.button,
        color: 'primary',
        disabled: disabled,
        onClick: onDeleteClick
      },
      !iconOnly && 'Delete',
      _react2.default.createElement(_Delete2.default, { className: (0, _classnames2.default)(classes.icon, { [classes.smallIcon]: iconOnly }) })
    );
  }
}

exports.Controls = Controls;
Controls.propTypes = {
  iconOnly: _propTypes2.default.bool.isRequired,
  classes: _propTypes2.default.object.isRequired,
  onDeleteClick: _propTypes2.default.func.isRequired,
  disabled: _propTypes2.default.bool.isRequired
};
const styles = theme => ({
  smallButton: {
    margin: 0,
    padding: 0,
    minWidth: 0,
    width: '30px'
  },
  button: {
    margin: theme.spacing.unit
  },
  icon: {
    marginLeft: theme.spacing.unit
  }
});

exports.default = (0, _styles.withStyles)(styles)(Controls);
