// @pie-ui/number-line/lib/number-line/graph/arrow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Arrow = Arrow;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const style = {
  root: {
    fill: 'var(--arrow-color, black)'
  }
};

function Arrow({ x, y, direction, classes, className }) {
  let transform = `translate(${x || 0},${y})`;

  if (direction && direction === 'right') {
    transform += ' rotate(180)';
  }

  const names = (0, _classnames2.default)(classes.root, className);
  return _react2.default.createElement('path', { d: 'm 0,0 8,-5 0,10 -8,-5', transform: transform, className: names });
}

Arrow.propTypes = {
  y: _propTypes2.default.number,
  x: _propTypes2.default.number,
  direction: _propTypes2.default.oneOf(['left', 'right']),
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string
};

Arrow.defaultProps = {
  y: 0,
  x: 0,
  direction: 'left'
};

exports.default = (0, _reactJss2.default)(style)(Arrow);
