// @pie-lib/math-input/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.removeBrackets = exports.addBrackets = exports.HorizontalKeypad = exports.EditableMathInput = undefined;

var _editableMathInput = require('./editable-math-input');

var _editableMathInput2 = _interopRequireDefault(_editableMathInput);

var _horizontalKeypad = require('./horizontal-keypad');

var _horizontalKeypad2 = _interopRequireDefault(_horizontalKeypad);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.EditableMathInput = _editableMathInput2.default;
exports.HorizontalKeypad = _horizontalKeypad2.default;


var addLeftBracket = function addLeftBracket(s) {
  return s.indexOf('\\(') === 0 ? s : '\\(' + s;
};
var addRightBracket = function addRightBracket(s) {
  return s.indexOf('\\)') === s.length - 2 ? s : s + '\\)';
};
var rmLeftBracket = function rmLeftBracket(s) {
  return s.indexOf('\\(') === 0 ? s.substring(2) : s;
};
var rmRightBracket = function rmRightBracket(s) {
  return s.indexOf('\\)') === s.length - 2 ? s.substring(0, s.length - 2) : s;
};

var addBrackets = exports.addBrackets = function addBrackets(s) {
  return addRightBracket(addLeftBracket(s));
};
var removeBrackets = exports.removeBrackets = function removeBrackets(s) {
  return rmRightBracket(rmLeftBracket(s));
};
