// @pie-lib/scoring-config/lib/scoring-config-rows.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Rows = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Button = require('@material-ui/core/Button');

var _Button2 = _interopRequireDefault(_Button);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _cloneDeep = require('lodash/cloneDeep');

var _cloneDeep2 = _interopRequireDefault(_cloneDeep);

var _styles = require('@material-ui/core/styles');

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _row = require('./row');

var _row2 = _interopRequireDefault(_row);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var log = (0, _debug2.default)('@pie-elements:scoring-config:scoring-config-row');

var Rows = exports.Rows = function (_React$Component) {
  _inherits(Rows, _React$Component);

  function Rows(props) {
    _classCallCheck(this, Rows);

    var _this = _possibleConstructorReturn(this, (Rows.__proto__ || Object.getPrototypeOf(Rows)).call(this, props));

    _this.addRow = _this.addRow.bind(_this);
    return _this;
  }

  _createClass(Rows, [{
    key: 'onRowChange',
    value: function onRowChange(index, row) {
      var partialScoring = this.props.partialScoring;

      var ps = (0, _cloneDeep2.default)(partialScoring);
      ps.splice(index, 1, row);
      this.props.onChange(ps);
    }
  }, {
    key: 'onDelete',
    value: function onDelete(index) {
      var partialScoring = this.props.partialScoring;

      var ps = (0, _cloneDeep2.default)(partialScoring);
      ps.splice(index, 1);
      this.props.onChange(ps);
    }
  }, {
    key: 'addRow',
    value: function addRow() {
      var _props = this.props,
          partialScoring = _props.partialScoring,
          onChange = _props.onChange;

      var ps = (0, _cloneDeep2.default)(partialScoring);
      ps.push({
        numberOfCorrect: 1,
        scorePercentage: 50
      });
      onChange(ps);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props2 = this.props,
          partialScoring = _props2.partialScoring,
          numberOfCorrectResponses = _props2.numberOfCorrectResponses,
          classes = _props2.classes;

      var maxAnswers = Math.max(1, numberOfCorrectResponses - 1);
      var canAddRow = partialScoring ? partialScoring.length < maxAnswers : true;

      log('partialScoring: ', partialScoring);

      return _react2.default.createElement(
        'div',
        null,
        partialScoring.map(function (row, index) {
          return _react2.default.createElement(_row2.default, _extends({}, row, {
            onRowChange: function onRowChange(row) {
              return _this2.onRowChange(index, row);
            },
            onDelete: function onDelete() {
              return _this2.onDelete(index);
            },
            maxAnswers: maxAnswers,
            deletable: true,
            key: index
          }));
        }),
        _react2.default.createElement(
          _Button2.default,
          {
            className: classes.addButton,
            disabled: !canAddRow,
            variant: 'raised',
            color: 'primary',
            onClick: this.addRow
          },
          partialScoring.length > 0 ? 'Add another scenario' : 'Add scenario'
        )
      );
    }
  }]);

  return Rows;
}(_react2.default.Component);

Rows.propTypes = {
  numberOfCorrectResponses: _propTypes2.default.number.isRequired,
  partialScoring: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    numberOfCorrect: _propTypes2.default.number.isRequired,
    scorePercentage: _propTypes2.default.number.isRequired
  })),
  onChange: _propTypes2.default.func.isRequired,
  classes: _propTypes2.default.object.isRequired
};
Rows.defaultProps = {
  partialScoring: []
};


var styles = function styles(theme) {
  return {
    addButton: {
      marginTop: theme.spacing.unit * 2
    }
  };
};

exports.default = (0, _styles.withStyles)(styles)(Rows);
