// @pie-ui/function-entry/lib/radical.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _styles = require('@material-ui/core/styles');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = () => ({
  radicalSymbol: {
    textDecoration: 'overline'
  }
});

class Radical extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          sqrt = _props.sqrt;


    return _react2.default.createElement(
      'span',
      null,
      '\u221A',
      _react2.default.createElement(
        'span',
        { className: classes.radicalSymbol },
        _react2.default.createElement(
          'i',
          null,
          sqrt
        ),
        '\xA0'
      )
    );
  }
}

Radical.propTypes = {
  classes: _propTypes2.default.object,
  sqrt: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string])
};
exports.default = (0, _styles.withStyles)(styles)(Radical);
