// @pie-element/placement-ordering-configure/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _pieConfigureEvents = require('@pie-framework/pie-configure-events');

var _main = require('./main');

var _main2 = _interopRequireDefault(_main);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class PlacementOrdering extends HTMLElement {
  constructor() {
    super();
    this.onModelChange = (model, resetSession) => {
      this._model = model;
      this.dispatchUpdate(resetSession);
    };

    this.insertImage = handler => {
      this.dispatchEvent(new _pieConfigureEvents.InsertImageEvent(handler));
    };

    this.deleteImage = (src, done) => {
      this.dispatchEvent(new _pieConfigureEvents.DeleteImageEvent(src, done));
    };
  }

  dispatchUpdate(reset) {
    const detail = { update: this._model, reset };
    this.dispatchEvent(new _pieConfigureEvents.ModelUpdatedEvent(this._model, reset));
  }

  set model(s) {
    this._model = s;
    this._rerender();
  }

  _rerender() {
    let element = _react2.default.createElement(_main2.default, {
      initialModel: this._model,
      onModelChange: this.onModelChange,
      imageSupport: {
        add: this.insertImage,
        delete: this.deleteImage
      }
    });
    _reactDom2.default.render(element, this);
  }
}
exports.default = PlacementOrdering;
