// @pie-ui/calculator/lib/main.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _draggableCalculator = require('./draggable-calculator');

var _draggableCalculator2 = _interopRequireDefault(_draggableCalculator);

var _calculatorIcon = require('./calculator-icon');

var _calculatorIcon2 = _interopRequireDefault(_calculatorIcon);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = {
  icon: {
    transition: 'fill 200ms',
    '&:hover': {
      fill: 'black'
    },
    cursor: 'pointer',
    verticalAlign: 'middle',
    fill: 'grey'
  },
  active: {
    fill: 'black'
  }
};

class Main extends _react2.default.Component {
  constructor(props) {
    super(props);

    this.onToggleShow = () => this.setState({ show: !this.state.show });

    this.onClickClose = () => this.setState({ show: false });

    this.state = { show: false };
  }

  render() {
    const show = this.state.show;
    const mode = this.props.model.mode;
    const classes = this.props.classes;

    return _react2.default.createElement(
      'div',
      null,
      _react2.default.createElement(_calculatorIcon2.default, {
        className: show ? (0, _classnames2.default)(classes.icon, classes.active) : classes.icon,
        onClick: () => this.onToggleShow()
      }),
      _react2.default.createElement(_draggableCalculator2.default, {
        mode: mode,
        show: show,
        onClose: this.onClickClose
      })
    );
  }
}

Main.propTypes = {
  model: _propTypes2.default.object.isRequired,
  classes: _propTypes2.default.object.isRequired
};
exports.default = (0, _styles.withStyles)(styles)(Main);
