// @pie-ui/function-entry/lib/fraction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _styles = require('@material-ui/core/styles');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = theme => ({
  fraction: {
    display: 'inline-block',
    position: 'relative',
    verticalAlign: 'middle',
    letterSpacing: '0.001em',
    textAlign: 'center',
    marginRight: theme.spacing.unit
  },
  fractionSpan: {
    display: 'block',
    padding: '0.1em'
  },
  fractionBottom: {
    borderTop: 'thin solid black'
  },
  fractionSymbol: {
    display: 'none'
  }
});

class Fraction extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          top = _props.top,
          bottom = _props.bottom;


    return _react2.default.createElement(
      'span',
      { className: classes.fraction },
      _react2.default.createElement(
        'span',
        null,
        _react2.default.createElement(
          'i',
          null,
          top
        )
      ),
      _react2.default.createElement(
        'span',
        { className: (0, _classnames2.default)(classes.fractionSpan, classes.fractionSymbol) },
        '/'
      ),
      _react2.default.createElement(
        'span',
        { className: (0, _classnames2.default)(classes.fractionSpan, classes.fractionBottom) },
        _react2.default.createElement(
          'i',
          null,
          bottom
        )
      )
    );
  }
}

Fraction.propTypes = {
  classes: _propTypes2.default.object,
  top: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),
  bottom: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string])
};
exports.default = (0, _styles.withStyles)(styles)(Fraction);
