// @pie-ui/function-entry/lib/hints.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _styles = require('@material-ui/core/styles');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Typography = require('@material-ui/core/Typography');

var _Typography2 = _interopRequireDefault(_Typography);

var _fraction = require('./fraction');

var _fraction2 = _interopRequireDefault(_fraction);

var _power = require('./power');

var _power2 = _interopRequireDefault(_power);

var _radical = require('./radical');

var _radical2 = _interopRequireDefault(_radical);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = theme => ({
  content: {
    paddingRight: theme.spacing.unit * 2
  }
});

class Hints extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          className = _props.className;


    return _react2.default.createElement(
      'ul',
      { className: className },
      _react2.default.createElement(
        _Typography2.default,
        { type: 'body1', className: classes.content },
        _react2.default.createElement(
          'li',
          null,
          'For 2 \u2219 2 enter 2*2'
        ),
        _react2.default.createElement(
          'li',
          null,
          'For 3y enter 3y or 3*y'
        ),
        _react2.default.createElement(
          'li',
          null,
          'For ',
          _react2.default.createElement(_fraction2.default, { top: 1, bottom: 'x' }),
          ' enter 1/x'
        ),
        _react2.default.createElement(
          'li',
          null,
          'For ',
          _react2.default.createElement(_fraction2.default, { top: 1, bottom: 'xy' }),
          ' enter 1/(x*y)'
        ),
        _react2.default.createElement(
          'li',
          null,
          'For ',
          _react2.default.createElement(_fraction2.default, { top: 2, bottom: 'x+3' }),
          ' enter 2/(x+3)'
        ),
        _react2.default.createElement(
          'li',
          null,
          'For ',
          _react2.default.createElement(_power2.default, { base: 'x', exponent: 'y' }),
          ' enter (x ^ y)'
        ),
        _react2.default.createElement(
          'li',
          null,
          'For ',
          _react2.default.createElement(_power2.default, { base: 'x', exponent: '2' }),
          ' enter (x ^ 2)'
        ),
        _react2.default.createElement(
          'li',
          null,
          'For 1 ',
          _react2.default.createElement(_fraction2.default, { top: 'x', bottom: 'y' }),
          ' enter 1 x/y'
        ),
        _react2.default.createElement(
          'li',
          null,
          'For ',
          _react2.default.createElement(_radical2.default, { sqrt: 'x' }),
          ' enter sqrt(x)'
        )
      )
    );
  }
}

Hints.propTypes = {
  className: _propTypes2.default.string,
  classes: _propTypes2.default.object
};
Hints.defaultProps = {
  className: ''
};
exports.default = (0, _styles.withStyles)(styles)(Hints);
