// @pie-ui/function-entry/lib/power.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _styles = require('@material-ui/core/styles');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = () => ({
  power: {
    fontSize: '12px',
    verticalAlign: '+50%',
    marginLeft: '2px'
  }
});

class Power extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          exponent = _props.exponent,
          base = _props.base;


    return _react2.default.createElement(
      'span',
      null,
      _react2.default.createElement(
        'i',
        null,
        base
      ),
      _react2.default.createElement(
        'span',
        { className: classes.power },
        exponent
      )
    );
  }
}

Power.propTypes = {
  classes: _propTypes2.default.object,
  exponent: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),
  base: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string])
};
exports.default = (0, _styles.withStyles)(styles)(Power);
