// @pie-ui/calculator/lib/draggable-calculator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DraggableCalculator = exports.CalculatorLayout = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _materialUiCalculator = require('@pie-framework/material-ui-calculator');

var _materialUiCalculator2 = _interopRequireDefault(_materialUiCalculator);

var _reactDraggable = require('react-draggable');

var _reactDraggable2 = _interopRequireDefault(_reactDraggable);

var _Typography = require('@material-ui/core/Typography');

var _Typography2 = _interopRequireDefault(_Typography);

var _Close = require('@material-ui/icons/Close');

var _Close2 = _interopRequireDefault(_Close);

var _IconButton = require('@material-ui/core/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _grey = require('@material-ui/core/colors/grey');

var _grey2 = _interopRequireDefault(_grey);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const styles = theme => {
  var _theme$palette = theme.palette;
  const secondary = _theme$palette.secondary,
        primary = _theme$palette.primary;


  const border = `solid 1px var(--pie-ui, ${_grey2.default[900]})`;

  return {
    baseLayout: {
      display: 'inline-block',
      border,
      zIndex: '10'
    },
    scientific: {
      minWidth: '515px'
    },
    closeIcon: {
      width: '24px',
      height: '24px',
      color: 'white'
    },
    title: {
      color: secondary.contrastText,
      flex: 1
    },
    titleBar: {
      cursor: 'move',
      position: 'relative',
      display: 'flex',
      alignItems: 'center',
      padding: theme.spacing.unit,
      backgroundColor: primary.light,
      borderBottom: border
    }
  };
};

class BaseLayout extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          onClose = _props.onClose,
          mode = _props.mode,
          className = _props.className;

    return _react2.default.createElement(
      'div',
      {
        className: (0, _classnames2.default)(classes.baseLayout, mode === 'scientific' && classes.scientific, className)
      },
      _react2.default.createElement(
        'div',
        { className: `handle ${classes.titleBar}` },
        _react2.default.createElement(
          _Typography2.default,
          { variant: 'subheading', className: classes.title },
          'Calculator'
        ),
        _react2.default.createElement(
          _IconButton2.default,
          {
            className: classes.closeIcon,
            onClick: onClose,
            'aria-label': 'Delete'
          },
          _react2.default.createElement(_Close2.default, null)
        )
      ),
      _react2.default.createElement(_materialUiCalculator2.default, { mode: mode })
    );
  }
}

BaseLayout.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string,
  onClose: _propTypes2.default.func.isRequired,
  mode: _propTypes2.default.oneOf(['basic', 'scientific'])
};
const CalculatorLayout = exports.CalculatorLayout = (0, _styles.withStyles)(styles)(BaseLayout);

class DraggableCalculator extends _react2.default.Component {

  constructor(props) {
    super(props);

    this.handleDrag = (e, ui) => {
      var _state$deltaPosition = this.state.deltaPosition;
      const x = _state$deltaPosition.x,
            y = _state$deltaPosition.y;

      this.setState({
        deltaPosition: {
          x: x + ui.deltaX,
          y: y + ui.deltaY
        }
      });
    };

    this.state = {
      deltaPosition: {
        x: 0,
        y: 0
      }
    };
  }

  render() {
    var _props2 = this.props;
    const mode = _props2.mode,
          show = _props2.show,
          onClose = _props2.onClose,
          classes = _props2.classes;
    var _state$deltaPosition2 = this.state.deltaPosition;
    const x = _state$deltaPosition2.x,
          y = _state$deltaPosition2.y;

    return show ? _react2.default.createElement(
      _reactDraggable2.default,
      {
        onDrag: this.handleDrag,
        defaultPosition: { x, y },
        handle: '.handle'
      },
      _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(CalculatorLayout, {
          onClose: onClose,
          mode: mode,
          className: classes.draggable
        })
      )
    ) : null;
  }
}

exports.DraggableCalculator = DraggableCalculator;
DraggableCalculator.propTypes = {
  mode: _propTypes2.default.oneOf(['basic', 'scientific']),
  show: _propTypes2.default.bool.isRequired,
  onClose: _propTypes2.default.func.isRequired,
  classes: _propTypes2.default.object.isRequired
};
exports.default = (0, _styles.withStyles)(() => ({
  draggable: {
    position: 'absolute'
  }
}))(DraggableCalculator);
