// @pie-lib/text-select/lib/tokenizer/controls.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Controls = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Button = require('@material-ui/core/Button');

var _Button2 = _interopRequireDefault(_Button);

var _styles = require('@material-ui/core/styles');

var _Switch = require('@material-ui/core/Switch');

var _Switch2 = _interopRequireDefault(_Switch);

var _FormControlLabel = require('@material-ui/core/FormControlLabel');

var _FormControlLabel2 = _interopRequireDefault(_FormControlLabel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Controls = exports.Controls = function (_React$Component) {
  _inherits(Controls, _React$Component);

  function Controls() {
    _classCallCheck(this, Controls);

    return _possibleConstructorReturn(this, (Controls.__proto__ || Object.getPrototypeOf(Controls)).apply(this, arguments));
  }

  _createClass(Controls, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          classes = _props.classes,
          onClear = _props.onClear,
          onWords = _props.onWords,
          onSentences = _props.onSentences,
          setCorrectMode = _props.setCorrectMode,
          onToggleCorrectMode = _props.onToggleCorrectMode;

      return _react2.default.createElement(
        'div',
        { className: classes.controls },
        _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(
            _Button2.default,
            {
              onClick: onWords,
              className: classes.button,
              size: 'small',
              color: 'primary',
              disabled: setCorrectMode
            },
            'Words'
          ),
          _react2.default.createElement(
            _Button2.default,
            {
              onClick: onSentences,
              className: classes.button,
              size: 'small',
              color: 'primary',
              disabled: setCorrectMode
            },
            'Sentences'
          ),
          _react2.default.createElement(
            _Button2.default,
            {
              className: classes.button,
              size: 'small',
              color: 'secondary',
              onClick: onClear,
              disabled: setCorrectMode
            },
            'Clear'
          )
        ),
        _react2.default.createElement(_FormControlLabel2.default, {
          control: _react2.default.createElement(_Switch2.default, { checked: setCorrectMode, onChange: onToggleCorrectMode }),
          label: 'Set correct answers'
        })
      );
    }
  }]);

  return Controls;
}(_react2.default.Component);

Controls.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  onClear: _propTypes2.default.func.isRequired,
  onWords: _propTypes2.default.func.isRequired,
  onSentences: _propTypes2.default.func.isRequired,
  setCorrectMode: _propTypes2.default.bool.isRequired,
  onToggleCorrectMode: _propTypes2.default.func.isRequired
};
Controls.defaultProps = {};
exports.default = (0, _styles.withStyles)(function (theme) {
  return {
    button: {
      marginRight: theme.spacing.unit
    },
    controls: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'space-between'
    }
  };
})(Controls);
