// @pie-lib/math-input/lib/styles.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buttonStyle = exports.OverrideIconButton = undefined;

var _IconButton = require('@material-ui/core/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OverrideIconButton = exports.OverrideIconButton = function OverrideIconButton(props) {
  return _react2.default.createElement(
    _IconButton2.default,
    {
      tabIndex: '-1',
      classes: { root: props.classes.root, label: props.classes.label }
    },
    props.children
  );
};

OverrideIconButton.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node]).isRequired
};

var buttonStyle = exports.buttonStyle = function buttonStyle() {
  return {
    root: {
      borderRadius: '0',
      marginRight: '5px',
      marginBottom: '0px',
      width: '100%'
    },
    label: {
      display: 'block'
    }
  };
};
