// @pie-ui/point-intercept/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Component = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _main = require('./main');

var _main2 = _interopRequireDefault(_main);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var log = (0, _debug2.default)('pie-ui:point-intercept');

exports.Component = _main2.default;

var PointIntercept = function (_HTMLElement) {
  _inherits(PointIntercept, _HTMLElement);

  function PointIntercept() {
    _classCallCheck(this, PointIntercept);

    return _possibleConstructorReturn(this, (PointIntercept.__proto__ || Object.getPrototypeOf(PointIntercept)).call(this));
  }

  _createClass(PointIntercept, [{
    key: 'sessionChanged',
    value: function sessionChanged(s) {
      this._session.points = s.points;
      log('session: ', this._session);
    }
  }, {
    key: 'connectedCallback',
    value: function connectedCallback() {
      this._render();
    }
  }, {
    key: '_render',
    value: function _render() {
      if (!this._model || !this._session) {
        return;
      }

      var el = _react2.default.createElement(_main2.default, {
        model: this._model,
        session: this._session,
        onSessionChange: this.sessionChanged.bind(this)
      });

      _reactDom2.default.render(el, this);
    }
  }, {
    key: 'model',
    set: function set(m) {
      this._model = m;
      this._render();
    }
  }, {
    key: 'session',
    set: function set(s) {
      this._session = s;
      this._render();
    }
  }]);

  return PointIntercept;
}(HTMLElement);

exports.default = PointIntercept;
