// @pie-element/categorize/configure/lib/design/buttons.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Divider = exports.DeleteButton = exports.AddButton = exports.RawAddButton = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Button = require('@material-ui/core/Button');

var _Button2 = _interopRequireDefault(_Button);

var _Divider = require('@material-ui/core/Divider');

var _Divider2 = _interopRequireDefault(_Divider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class RawAddButton extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          className = _props.className,
          label = _props.label,
          onClick = _props.onClick,
          disabled = _props.disabled;

    return _react2.default.createElement(
      _Button2.default,
      {
        onClick: onClick,
        disabled: disabled,
        className: (0, _classnames2.default)(classes.addButton, className),
        size: 'small',
        variant: 'raised',
        color: 'primary'
      },
      label
    );
  }
}
exports.RawAddButton = RawAddButton;
RawAddButton.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string,
  label: _propTypes2.default.string,
  onClick: _propTypes2.default.func,
  disabled: _propTypes2.default.bool
};
RawAddButton.defaultProps = {
  label: 'Add'
};
const styles = theme => ({
  addButton: {
    height: theme.spacing.unit * 4
  }
});

const AddButton = (0, _styles.withStyles)(styles)(RawAddButton);

const DeleteButton = (0, _styles.withStyles)(theme => ({
  deleteButton: {
    margin: 0,
    padding: 0
  }
}))(({ classes, label, onClick }) => _react2.default.createElement(
  _Button2.default,
  {
    className: classes.deleteButton,
    onClick: onClick,
    size: 'small',
    color: 'primary'
  },
  label
));

const Divider = (0, _styles.withStyles)(theme => ({
  divider: {
    marginTop: 2 * theme.spacing.unit,
    marginBottom: 2 * theme.spacing.unit
  }
}))(({ classes }) => _react2.default.createElement(_Divider2.default, { className: classes.divider }));

exports.AddButton = AddButton;
exports.DeleteButton = DeleteButton;
exports.Divider = Divider;
