// @pie-element/categorize/configure/lib/design/header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Header = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Typography = require('@material-ui/core/Typography');

var _Typography2 = _interopRequireDefault(_Typography);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _buttons = require('./buttons');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Header extends _react2.default.Component {
  render() {
    var _props = this.props;
    const classes = _props.classes,
          className = _props.className,
          onAdd = _props.onAdd,
          label = _props.label,
          buttonLabel = _props.buttonLabel;

    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(classes.header, className) },
      _react2.default.createElement(
        _Typography2.default,
        { variant: 'title', className: classes.title },
        label
      ),
      _react2.default.createElement(_buttons.AddButton, { onClick: onAdd, label: buttonLabel })
    );
  }
}
exports.Header = Header;
Header.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string,
  buttonLabel: _propTypes2.default.string,
  onAdd: _propTypes2.default.func.isRequired,
  label: _propTypes2.default.string.isRequired
};
Header.defaultProps = {};
const styles = theme => ({
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    marginBottom: 2 * theme.spacing.unit
  }
});
exports.default = (0, _styles.withStyles)(styles)(Header);
