// @pie-element/categorize/configure/lib/design/input-header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputHeader = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _editableHtml = require('@pie-lib/editable-html');

var _editableHtml2 = _interopRequireDefault(_editableHtml);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class InputHeader extends _react2.default.Component {
  render() {
    var _props = this.props;
    const onChange = _props.onChange,
          label = _props.label,
          classes = _props.classes,
          className = _props.className,
          imageSupport = _props.imageSupport;


    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(classes.inputHeader, className) },
      _react2.default.createElement(_editableHtml2.default, {
        imageSupport: imageSupport,
        autoWidthToolbar: true,
        label: 'label',
        markup: label,
        onChange: onChange,
        className: classes.editor
      })
    );
  }
}
exports.InputHeader = InputHeader;
InputHeader.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string,
  label: _propTypes2.default.string,
  onChange: _propTypes2.default.func,
  onDelete: _propTypes2.default.func,
  imageSupport: _propTypes2.default.shape({
    add: _propTypes2.default.func.isRequired,
    delete: _propTypes2.default.func.isRequired
  })
};
InputHeader.defaultProps = {};
const styles = theme => ({
  editor: {
    flex: '1',
    paddingBottom: theme.spacing.unit
  },
  iconButtonRoot: {
    width: 'auto',
    height: 'auto'
  },
  inputHeader: {
    display: 'flex',
    justifyContent: 'space-between'
  }
});
exports.default = (0, _styles.withStyles)(styles)(InputHeader);
