// @pie-lib/editable-html/lib/plugins/math/component.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MathComponent = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _slate = require('slate');

var _mathInput = require('@pie-lib/math-input');

var _inputWrapper = require('./input-wrapper');

var _inputWrapper2 = _interopRequireDefault(_inputWrapper);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _theme = require('../../theme');

var _slatePropTypes = require('slate-prop-types');

var _slatePropTypes2 = _interopRequireDefault(_slatePropTypes);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var log = (0, _debug2.default)('editable-html:plugins:math:component');

var MathComponent = exports.MathComponent = function (_React$Component) {
  _inherits(MathComponent, _React$Component);

  function MathComponent(props) {
    _classCallCheck(this, MathComponent);

    var _this = _possibleConstructorReturn(this, (MathComponent.__proto__ || Object.getPrototypeOf(MathComponent)).call(this, props));

    _this.onFocus = function (event) {
      log('[onFocus]');
      if (_this.props.onFocus) {
        _this.props.onFocus(event);
      }
    };

    _this.onBlur = function (event) {
      log('[onBlur]');
      if (_this.props.onBlur) {
        _this.props.onBlur(event);
      }
    };

    _this.onChange = function (latex) {
      log('[onChange]', latex);
      var _this$props = _this.props,
          node = _this$props.node,
          editor = _this$props.editor;
      var key = node.key;

      var data = _slate.Data.create({ latex: latex, editing: true });
      var change = editor.value.change().setNodeByKey(key, { data: data });
      editor.onChange(change);
    };

    _this.onClick = function (event) {
      log('[onClick] preventDefault and stopPropagation');
      if (_this.props.onClick) {
        event.preventDefault();
        event.stopPropagation();
        _this.props.onClick();
      }
    };

    _this.onDeleteClick = function (event) {
      log('delete click');
      event.preventDefault();
      event.stopPropagation();

      var _this$props2 = _this.props,
          node = _this$props2.node,
          editor = _this$props2.editor;


      var change = editor.value.change().moveNodeByKey(node.key);
      editor.onChange(change);
    };

    _this.onMathChange = function (latex) {
      log('[onMathChange]', latex);
      var _this$props3 = _this.props,
          node = _this$props3.node,
          editor = _this$props3.editor;

      var data = _slate.Data.create({ latex: latex });
      editor.change(function (c) {
        return c.setNodeByKey(node.key, { data: data });
      });
    };

    _this.state = {
      disableBlur: false
    };
    return _this;
  }

  _createClass(MathComponent, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      log('this.wrapper: ', this.wrapper);
      var _props = this.props,
          node = _props.node,
          editor = _props.editor;

      var mathChange = node.data.get('change');

      if (mathChange) {
        var latex = this.wrapper.change(mathChange);

        var data = node.data.toObject();
        delete data.change;
        data.latex = latex;

        log('[componentDidUpdate] new latex: ', data.latex);

        editor.change(function (c) {
          c.setNodeByKey(node.key, { data: data });
        });
      }
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      log('this.wrapper: ', this.wrapper);
    }
  }, {
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps) {
      var nextNode = nextProps.node;
      var node = this.props.node;

      if (nextNode.data.equals(node.data)) {
        return false;
      }

      if (nextNode.data.get('change') === undefined && node.data.get('change') !== undefined) {
        return false;
      }
      return true;
    }
  }, {
    key: 'blur',
    value: function blur() {
      log('[blur]');
      this.wrapper && this.wrapper.blur();
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props2 = this.props,
          node = _props2.node,
          classes = _props2.classes,
          editor = _props2.editor;

      log('[render] >> node', node, 'editor: ', editor, 'value: ', editor.value, editor.value.isFocused);

      var latex = node.data.get('latex');
      var names = (0, _classnames2.default)(classes.root);
      var cleanLatex = (0, _mathInput.removeBrackets)(latex);

      return _react2.default.createElement(
        'div',
        { className: names },
        _react2.default.createElement(_inputWrapper2.default, {
          ref: function ref(r) {
            return _this2.wrapper = r;
          },
          latex: cleanLatex,
          editing: true,
          onClick: this.onClick,
          onChange: this.onMathChange,
          onFocus: this.onFocus,
          onBlur: this.onBlur
        })
      );
    }
  }]);

  return MathComponent;
}(_react2.default.Component);

MathComponent.propTypes = {
  node: _slatePropTypes2.default.node.isRequired,
  editor: _propTypes2.default.shape({
    value: _slatePropTypes2.default.value.isRequired
  }),
  onFocus: _propTypes2.default.func,
  onBlur: _propTypes2.default.func,
  onClick: _propTypes2.default.func.isRequired,
  classes: _propTypes2.default.object.isRequired
};


var styles = {
  root: {
    display: 'inline-flex',
    alignItems: 'center',
    '& > .mq-editable-field': {
      border: 'solid 1px lightgrey'
    },
    '& > .mq-focused': {
      outline: 'none',
      boxShadow: 'none',
      border: 'solid 1px ' + _theme.primary,
      borderRadius: '0px'
    }
  },
  selected: {
    border: 'solid 1px red'
  }
};

exports.default = (0, _reactJss2.default)(styles)(MathComponent);
