// @pie-framework/material-ui-calculator/lib/pad.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pad = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styles = require('@material-ui/core/styles');

var _IconButton = require('@material-ui/core/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _colors = require('./colors');

var colors = _interopRequireWildcard(_colors);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Pad = exports.Pad = function Pad(props) {
  var label = props.label,
      classes = props.classes,
      theme = props.theme,
      style = props.style,
      value = props.value,
      onClick = props.onClick;


  var names = (0, _classnames2.default)(classes.pad, theme && theme.root);

  var handleClick = function handleClick() {
    onClick(value);
  };

  return _react2.default.createElement(
    'div',
    { style: style, className: names },
    _react2.default.createElement(
      _IconButton2.default,
      {
        tabIndex: '-1',
        className: classes.button,
        onClick: handleClick
      },
      _react2.default.createElement('div', { dangerouslySetInnerHTML: { __html: label } })
    )
  );
};

Pad.propTypes = {
  label: _propTypes2.default.string.isRequired,
  classes: _propTypes2.default.object.isRequired,
  theme: _propTypes2.default.object,
  style: _propTypes2.default.object,
  value: _propTypes2.default.string.isRequired,
  onClick: _propTypes2.default.func.isRequired
};

exports.default = (0, _styles.withStyles)(function () {
  return {
    pad: {
      backgroundColor: colors.primary.main
    },
    button: {
      width: '100%'
    }
  };
})(Pad);
