// @pie-framework/material-ui-calculator/lib/angle-mode.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _colors = require('./colors');

var colors = _interopRequireWildcard(_colors);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Angle = (0, _styles.withStyles)(function (theme) {
  return {
    mode: {
      cursor: 'pointer',
      fontFamily: theme.typography.fontFamily,
      textTransform: 'uppercase',
      color: colors.disabled
    },
    active: {
      color: colors.primary.main
    }
  };
})(function (_ref) {
  var classes = _ref.classes,
      onChange = _ref.onChange,
      type = _ref.type,
      angleMode = _ref.angleMode;
  return _react2.default.createElement(
    'div',
    {
      onClick: function onClick() {
        return onChange(type);
      },
      className: (0, _classnames2.default)(classes.mode, angleMode === type && classes.active)
    },
    type
  );
});

var AngleMode = function AngleMode(_ref2) {
  var className = _ref2.className,
      angleMode = _ref2.angleMode,
      onChange = _ref2.onChange;
  return _react2.default.createElement(
    'div',
    { className: className },
    _react2.default.createElement(Angle, { type: 'deg', onChange: onChange, angleMode: angleMode }),
    _react2.default.createElement(Angle, { type: 'rad', onChange: onChange, angleMode: angleMode })
  );
};

AngleMode.propTypes = {
  className: _propTypes2.default.string,
  angleMode: _propTypes2.default.oneOf(['deg', 'rad']),
  onChange: _propTypes2.default.func.isRequired
};

exports.default = AngleMode;
