// @pie-lib/editable-html/lib/plugins/toolbar/done-button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DoneButton = exports.RawDoneButton = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _IconButton = require('@material-ui/core/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _Check = require('@material-ui/icons/Check');

var _Check2 = _interopRequireDefault(_Check);

var _styles = require('@material-ui/core/styles');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RawDoneButton = exports.RawDoneButton = function RawDoneButton(_ref) {
  var classes = _ref.classes,
      onClick = _ref.onClick;
  return _react2.default.createElement(
    _IconButton2.default,
    {
      'aria-label': 'Done',
      className: classes.iconRoot,
      onClick: onClick,
      classes: {
        label: classes.label,
        root: classes.iconRoot
      }
    },
    _react2.default.createElement(_Check2.default, null)
  );
};

RawDoneButton.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  onClick: _propTypes2.default.func
};

var styles = {
  iconRoot: {
    verticalAlign: 'top',
    width: '28px',
    height: '28px',
    color: 'var(--editable-html-toolbar-check, #00bb00)'
  }
};
var DoneButton = exports.DoneButton = (0, _styles.withStyles)(styles)(RawDoneButton);
