// @pie-lib/charting/lib/arrow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Arrow = Arrow;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var style = {
  root: {
    fill: 'var(--arrow-color, black)'
  }
};

function Arrow(_ref) {
  var x = _ref.x,
      y = _ref.y,
      direction = _ref.direction,
      classes = _ref.classes,
      className = _ref.className;

  var transform = 'translate(' + (x || 0) + ',' + y + ')';

  if (direction) {
    if (direction === 'right') {
      transform += ' rotate(180)';
    }
    if (direction === 'up') {
      transform += ' rotate(90)';
    }
    if (direction === 'down') {
      transform += ' rotate(270)';
    }
  }

  var names = (0, _classnames2.default)(classes.root, className);
  return _react2.default.createElement('path', { d: 'm 0,0 8,-5 0,10 -8,-5', transform: transform, className: names });
}

Arrow.propTypes = {
  y: _propTypes2.default.number,
  x: _propTypes2.default.number,
  direction: _propTypes2.default.oneOf(['left', 'right', 'up', 'down']),
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string
};

Arrow.defaultProps = {
  y: 0,
  x: 0,
  direction: 'left'
};

exports.default = (0, _reactJss2.default)(style)(Arrow);
