// @pie-lib/charting/lib/color.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var C = function () {
  function C(name, main, text) {
    _classCallCheck(this, C);

    this.name = name;
    this.raw = {
      main: main,
      text: text
    };
  }

  _createClass(C, [{
    key: 'main',
    get: function get() {
      return 'var(--pie-' + this.name + ', ' + this.raw.main + ')';
    }
  }, {
    key: 'text',
    get: function get() {
      return 'var(--pie-' + this.name + '-text, ' + this.raw.text + ')';
    }
  }]);

  return C;
}();

var correct = exports.correct = new C('correct', '#4caf50', '#000000');

var incorrect = exports.incorrect = new C('incorrect', '#ff6f00', '#000000');

var primary = exports.primary = new C('primary', '#2196f3', '#ffffff');

var secondary = exports.secondary = new C('secondary', 'rgb(225, 0, 80)', '#ffffff');

var disabled = exports.disabled = new C('disabled', '#9e9e9e', '#000000');
