// @pie-element/point-intercept/configure/lib/add-point.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AddPointButton = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Button = require('@material-ui/core/Button');

var _Button2 = _interopRequireDefault(_Button);

var _styles = require('@material-ui/core/styles');

var _Add = require('@material-ui/icons/Add');

var _Add2 = _interopRequireDefault(_Add);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class AddPointButton extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          onAddClick = _props.onAddClick;


    return _react2.default.createElement(
      _Button2.default,
      { className: classes.button, disabled: false, onClick: onAddClick },
      _react2.default.createElement(_Add2.default, null),
      'Add Another Point'
    );
  }
}

exports.AddPointButton = AddPointButton;
AddPointButton.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  onAddClick: _propTypes2.default.func.isRequired
};
const styles = theme => ({
  button: {
    margin: 0,
    marginBottom: theme.spacing.unit * 2
  }
});

exports.default = (0, _styles.withStyles)(styles)(AddPointButton);
