// @pie-element/point-intercept/configure/lib/delete.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeleteControl = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Button = require('@material-ui/core/Button');

var _Button2 = _interopRequireDefault(_Button);

var _styles = require('@material-ui/core/styles');

var _Delete = require('@material-ui/icons/Delete');

var _Delete2 = _interopRequireDefault(_Delete);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class DeleteControl extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          disabled = _props.disabled,
          onDeleteClick = _props.onDeleteClick;


    return _react2.default.createElement(
      _Button2.default,
      {
        className: classes.button,
        disabled: disabled,
        onClick: onDeleteClick
      },
      _react2.default.createElement(_Delete2.default, null)
    );
  }
}

exports.DeleteControl = DeleteControl;
DeleteControl.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  onDeleteClick: _propTypes2.default.func.isRequired,
  disabled: _propTypes2.default.bool.isRequired
};
const styles = () => ({
  button: {
    margin: 0
  }
});

exports.default = (0, _styles.withStyles)(styles)(DeleteControl);
