// @pie-element/multiple-choice/configure/lib/choice-type.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KeyType = exports.ChoiceType = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _configUi = require('@pie-lib/config-ui');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const ChoiceType = exports.ChoiceType = props => {
  let choiceProps = {
    header: props.header,
    defaultSelected: 'radio',
    value: props.value,
    onChange: props.onChange,
    one: {
      label: 'Radio',
      value: 'radio'
    },
    two: {
      label: 'Checkbox',
      value: 'checkbox'
    }
  };
  return _react2.default.createElement(_configUi.TwoChoice, choiceProps);
};

const KeyType = exports.KeyType = props => {
  let choiceProps = {
    header: props.header,
    defaultSelected: 'numbers',
    value: props.value,
    onChange: props.onChange,
    one: {
      label: 'Numbers',
      value: 'numbers'
    },
    two: {
      label: 'Letters',
      value: 'letters'
    }
  };
  return _react2.default.createElement(_configUi.TwoChoice, choiceProps);
};

KeyType.propTypes = ChoiceType.propTypes = {
  value: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func.isRequired
};
