// @pie-element/multiple-choice/configure/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _pieConfigureEvents = require('@pie-framework/pie-configure-events');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _root = require('./root');

var _root2 = _interopRequireDefault(_root);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _configUi = require('@pie-lib/config-ui');

var _defaults = require('lodash/defaults');

var _defaults2 = _interopRequireDefault(_defaults);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const log = (0, _debug2.default)('multiple-choice:configure');

const prepareCustomizationObject = model => {
  const defaultValues = {
    promptLabel: 'Prompt',
    responseTypeLabel: 'Response Type',
    choicesLabel: 'Choice Labels',
    addChoiceButtonLabel: 'Add a choice',
    enableSelectChoiceLabel: true,
    enableSelectResponseType: true,
    enableAddChoice: true,
    enableAddFeedBack: true,
    enableDeleteChoice: true
  };

  return _extends({}, model, {
    configure: (0, _defaults2.default)(model.configure, defaultValues)
  });
};

exports.default = class extends HTMLElement {
  constructor() {
    super();
    this.onModelChanged = this.onModelChanged.bind(this);
  }

  set model(s) {
    this._model = prepareCustomizationObject(_configUi.choiceUtils.normalizeChoices(s));
    this._render();
  }

  dispatchModelUpdated(reset) {
    const resetValue = !!reset;

    this.dispatchEvent(new _pieConfigureEvents.ModelUpdatedEvent(this._model, resetValue));
  }

  onModelChanged(m, reset) {
    this._model = m;
    this.dispatchModelUpdated(reset);
  }

  /**
   *
   * @param {done, progress, file} handler
   */
  insertImage(handler) {
    this.dispatchEvent(new _pieConfigureEvents.InsertImageEvent(handler));
  }

  onDeleteImage(src, done) {
    this.dispatchEvent(new _pieConfigureEvents.DeleteImageEvent(src, done));
  }

  _render() {
    log('_render');
    let element = _react2.default.createElement(_root2.default, {
      model: this._model,
      onModelChanged: this.onModelChanged,
      imageSupport: {
        add: this.insertImage.bind(this),
        delete: this.onDeleteImage.bind(this)
      }
    });
    _reactDom2.default.render(element, this);
  }
};
