// @pie-element/multiple-choice/configure/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _pieConfigureEvents = require('@pie-framework/pie-configure-events');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _root = require('./root');

var _root2 = _interopRequireDefault(_root);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _configUi = require('@pie-lib/config-ui');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const log = (0, _debug2.default)('multiple-choice:configure');

exports.default = class extends HTMLElement {
  constructor() {
    super();
    this.onModelChanged = this.onModelChanged.bind(this);
  }

  set model(s) {
    this._model = _configUi.choiceUtils.normalizeChoices(s);
    this._render();
  }

  dispatchModelUpdated() {
    this.dispatchEvent(new _pieConfigureEvents.ModelUpdatedEvent(this._model, false));
  }

  onModelChanged(m) {
    this._model = m;
    this.dispatchModelUpdated();
  }

  /**
   *
   * @param {done, progress, file} handler
   */
  insertImage(handler) {
    this.dispatchEvent(new _pieConfigureEvents.InsertImageEvent(handler));
  }

  onDeleteImage(src, done) {
    this.dispatchEvent(new _pieConfigureEvents.DeleteImageEvent(src, done));
  }

  _render() {
    log('_render');
    let element = _react2.default.createElement(_root2.default, {
      model: this._model,
      onModelChanged: this.onModelChanged,
      imageSupport: {
        add: this.insertImage.bind(this),
        delete: this.onDeleteImage.bind(this)
      }
    });
    _reactDom2.default.render(element, this);
  }
};
