// @pie-lib/config-ui/lib/input-container.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _InputLabel = require('@material-ui/core/InputLabel');

var _InputLabel2 = _interopRequireDefault(_InputLabel);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styles = require('@material-ui/core/styles');

var _FormControl = require('@material-ui/core/FormControl');

var _FormControl2 = _interopRequireDefault(_FormControl);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RawInputContainer = function RawInputContainer(props) {
  var label = props.label,
      className = props.className,
      children = props.children,
      classes = props.classes;

  var names = (0, _classnames2.default)(classes.formControl, className);
  return _react2.default.createElement(
    _FormControl2.default,
    { className: names },
    _react2.default.createElement(
      _InputLabel2.default,
      { className: classes.label, shrink: true },
      label
    ),
    children
  );
};

RawInputContainer.propTypes = {
  label: _propTypes2.default.string.isRequired,
  className: _propTypes2.default.string,
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node]).isRequired,
  classes: _propTypes2.default.object.isRequired
};

exports.default = (0, _styles.withStyles)(function (theme) {
  return {
    formControl: {
      marginLeft: 0,
      marginRight: theme.spacing.unit,
      paddingBottom: theme.spacing.unit,
      flex: '1 0 auto',
      minWidth: theme.spacing.unit * 4
    },
    label: {
      whiteSpace: 'nowrap'
    }
  };
})(RawInputContainer);
