// @pie-lib/drag/lib/placeholder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PlaceHolder = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _grey = require('@material-ui/core/colors/grey');

var _grey2 = _interopRequireDefault(_grey);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PlaceHolder = exports.PlaceHolder = function PlaceHolder(_ref) {
  var children = _ref.children,
      classes = _ref.classes,
      className = _ref.className,
      isOver = _ref.isOver,
      type = _ref.type,
      grid = _ref.grid,
      disabled = _ref.disabled;

  var names = (0, _classnames2.default)(classes.placeholder, disabled && classes.disabled, isOver && classes.over, classes[type], className);

  var style = {};

  if (grid && grid.columns) {
    style.gridTemplateColumns = 'repeat(' + grid.columns + ', 1fr)';
  }
  if (grid && grid.rows) {
    style.gridTemplateRows = 'repeat(' + grid.rows + ', 1fr)';
  }
  return _react2.default.createElement(
    'div',
    { style: style, className: names },
    children
  );
};

PlaceHolder.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  grid: _propTypes2.default.shape({
    columns: _propTypes2.default.number,
    rows: _propTypes2.default.number
  }),
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node]).isRequired,
  className: _propTypes2.default.string,
  isOver: _propTypes2.default.bool,
  index: _propTypes2.default.number,
  type: _propTypes2.default.string,
  disabled: _propTypes2.default.bool
};

var styles = function styles(theme) {
  return {
    placeholder: {
      width: '100%',
      height: '100%',
      background: '#EEEEEE',
      border: '1px solid #D1D1D1',
      transition: 'background-color 200ms linear, border-color 200ms linear',
      boxSizing: 'border-box',
      display: 'grid',
      gridRowGap: theme.spacing.unit + 'px',
      gridColumnGap: theme.spacing.unit + 'px',
      padding: theme.spacing.unit * 1
    },
    disabled: {
      boxShadow: 'none',
      background: 'white'
    },
    over: {
      border: '1px solid ' + _grey2.default[500],
      backgroundColor: '' + _grey2.default[300]
    }
  };
};

exports.default = (0, _styles.withStyles)(styles)(PlaceHolder);
