// @pie-element/select-text-configure/lib/main.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Main = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _design = require('./design');

var _design2 = _interopRequireDefault(_design);

var _scoringConfig = require('@pie-lib/scoring-config');

var _scoringConfig2 = _interopRequireDefault(_scoringConfig);

var _configUi = require('@pie-lib/config-ui');

var _styles = require('@material-ui/core/styles');

var _cloneDeep = require('lodash/cloneDeep');

var _cloneDeep2 = _interopRequireDefault(_cloneDeep);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Main extends _react2.default.Component {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.changePartialScoring = partialScoring => {
      var _props = this.props;
      const onChange = _props.onChange,
            model = _props.model;

      const update = (0, _cloneDeep2.default)(model);
      update.partialScoring = partialScoring;
      onChange(update);
    }, _temp;
  }

  render() {
    var _props2 = this.props;
    const model = _props2.model,
          onChange = _props2.onChange,
          classes = _props2.classes;


    return _react2.default.createElement(
      _configUi.Tabs,
      null,
      _react2.default.createElement(_design2.default, {
        className: classes.container,
        title: 'Design',
        model: model,
        onChange: onChange
      }),
      _react2.default.createElement(_scoringConfig2.default, {
        title: 'Scoring',
        numberOfCorrectResponses: model.tokens.filter(t => t.correct === true).length,
        partialScoring: model.partialScoring,
        onChange: this.changePartialScoring
      })
    );
  }
}
exports.Main = Main;
Main.propTypes = {
  model: _propTypes2.default.object.isRequired,
  classes: _propTypes2.default.object.isRequired,
  onChange: _propTypes2.default.func.isRequired
};
Main.defaultProps = {};
exports.default = (0, _styles.withStyles)(theme => ({
  container: {
    paddingTop: theme.spacing.unit
  }
}))(Main);
