// @pie-element/categorize-configure/lib/scoring/grid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Grid = exports.Heading = exports.Cell = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _grey = require('@material-ui/core/colors/grey');

var _grey2 = _interopRequireDefault(_grey);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Cell = exports.Cell = (0, _styles.withStyles)(theme => ({
  cell: {
    width: '100%',
    height: '100%',
    backgroundColor: 'white',
    alignSelf: 'center',
    borderBottom: `solid 1px ${_grey2.default[300]}`,
    display: 'flex',
    alignItems: 'center'
  }
}))(({ classes, children, html }) => html ? _react2.default.createElement('div', {
  className: classes.cell,
  dangerouslySetInnerHTML: { __html: html }
}) : _react2.default.createElement(
  'div',
  { className: classes.cell },
  children
));

Cell.propTypes = {
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node]),
  html: _propTypes2.default.string
};
const Heading = exports.Heading = (0, _styles.withStyles)(theme => ({
  heading: {
    fontWeight: 'bold',
    paddingBottom: 0,
    marginBottom: theme.spacing.unit
  }
}))(({ classes, children }) => _react2.default.createElement(
  Cell,
  null,
  _react2.default.createElement(
    'div',
    { className: classes.heading },
    children
  )
));

class Grid extends _react2.default.Component {
  render() {
    var _props = this.props;
    const classes = _props.classes,
          className = _props.className,
          children = _props.children,
          headings = _props.headings,
          size = _props.size;

    const length = Array.isArray(headings) ? headings.length : 1;
    const columns = size ? size : `repeat(${length}, 1fr)`;
    const style = {
      gridTemplateColumns: columns
    };

    return _react2.default.createElement(
      'div',
      { style: style, className: (0, _classnames2.default)(classes.grid, className) },
      headings.map((h, index) => _react2.default.createElement(
        Heading,
        { key: index },
        h
      )),
      children
    );
  }
}
exports.Grid = Grid;
Grid.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string,
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node]).isRequired,
  headings: _propTypes2.default.arrayOf(_propTypes2.default.string),
  size: _propTypes2.default.string
};
Grid.defaultProps = {};
const styles = () => ({
  grid: {
    display: 'grid',
    backgroundColor: 'red',
    gridRowGap: '1px',
    gridColumnGap: 0
  }
});
exports.default = (0, _styles.withStyles)(styles)(Grid);
