// @pie-element/match/controller/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.model = model;

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _feedback = require('@pie-lib/feedback');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var log = (0, _debug2.default)('@pie-element:graph-lines:controller');

var getResponseCorrectness = function getResponseCorrectness(model, answers) {
  var allowPartialScores = model.allowPartialScoring;
  var partialScores = model.partialScoring;
  var rows = model.rows;
  var correctAnswers = 0;

  if (!answers || Object.keys(answers).length === 0) {
    return {
      correctness: 'empty',
      score: 0
    };
  }

  rows.forEach(function (row) {
    var isCorrectAnswer = answers[row.id] && JSON.stringify(answers[row.id]) === JSON.stringify(row.values);

    if (isCorrectAnswer) {
      correctAnswers += 1;
    }
  });

  if (rows.length === correctAnswers) {
    return { correctness: 'correct', score: '100%' };
  } else if (correctAnswers === 0) {
    return { correctness: 'incorrect', score: '0%' };
  } else if (allowPartialScores && partialScores && partialScores.length) {
    return {
      correctness: 'partial',
      score: ((partialScores.find(function (partialScore) {
        return partialScore.numberOfCorrect === correctAnswers;
      }) || {}).scorePercentage || 0) + '%'
    };
  }

  return { correctness: 'incorrect', score: '0%' };
};

function model(question, session, env) {
  return new Promise(function (resolve) {
    var getCorrectness = function getCorrectness() {
      if (env.mode === 'evaluate') {
        if (!session.answers || Object.keys(session.answers).length === 0) {
          return {
            correctness: 'unanswered',
            score: '0%'
          };
        }

        return getResponseCorrectness(question, session.answers);
      }
    };

    var correctInfo = getCorrectness();
    var correctResponse = {};

    question.rows.forEach(function (row) {
      correctResponse[row.id] = row.values;
    });

    var fb = env.mode === 'evaluate' ? (0, _feedback.getFeedbackForCorrectness)(correctInfo.correctness, question.feedback) : Promise.resolve(undefined);

    fb.then(function (feedback) {
      var base = {
        config: question,
        correctness: correctInfo,
        feedback: feedback,
        disabled: env.mode !== 'gather',
        view: env.mode === 'view'
      };

      var out = Object.assign(base, {
        correctResponse: env.mode === 'evaluate' ? correctResponse : undefined
      });
      log('out: ', out);
      resolve(out);
    });
  });
}
