// @pie-ui/categorize/lib/categorize/categories.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Categories = exports.CategoryType = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _category = require('./category');

var _category2 = _interopRequireDefault(_category);

var _gridContent = require('./grid-content');

var _gridContent2 = _interopRequireDefault(_gridContent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.CategoryType = _category.CategoryType;
class Categories extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          categories = _props.categories,
          config = _props.config,
          disabled = _props.disabled,
          _onDropChoice = _props.onDropChoice,
          onRemoveChoice = _props.onRemoveChoice,
          grid = _props.grid;


    return _react2.default.createElement(
      _gridContent2.default,
      { columns: config.columns, className: classes.categories },
      categories.map((c, index) => _react2.default.createElement(_category2.default, _extends({
        grid: grid,
        onDropChoice: h => _onDropChoice(c.id, h),
        onRemoveChoice: onRemoveChoice,
        disabled: disabled,
        className: classes.category,
        key: index
      }, c)))
    );
  }
}

exports.Categories = Categories;
Categories.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  categories: _propTypes2.default.arrayOf(_propTypes2.default.shape(_category.CategoryType)),
  config: _propTypes2.default.shape({
    columns: _propTypes2.default.number.isRequired
  }),
  disabled: _propTypes2.default.bool,
  onDropChoice: _propTypes2.default.func.isRequired,
  onRemoveChoice: _propTypes2.default.func.isRequired,
  grid: _propTypes2.default.object
};
Categories.defaultProps = {
  config: {
    columns: 4
  }
};
const styles = () => ({
  categories: {
    flex: 1
  }
});
exports.default = (0, _styles.withStyles)(styles)(Categories);
