// @pie-ui/categorize/lib/categorize/grid-content.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridContent = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class GridContent extends _react2.default.Component {

  render() {
    var _props = this.props;
    const classes = _props.classes,
          className = _props.className,
          children = _props.children,
          columns = _props.columns,
          extraStyle = _props.extraStyle;

    const style = _extends({
      gridTemplateColumns: `repeat(${columns}, 1fr)`
    }, extraStyle);

    return _react2.default.createElement(
      'div',
      { style: style, className: (0, _classnames2.default)(classes.gridContent, className) },
      children
    );
  }
}

exports.GridContent = GridContent;
GridContent.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string,
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node]).isRequired,
  columns: _propTypes2.default.number,
  extraStyle: _propTypes2.default.object
};
GridContent.defaultProps = {
  columns: 2
};
const styles = theme => ({
  gridContent: {
    display: 'grid',
    columnGap: `${theme.spacing.unit}px`,
    gridColumnGap: `${theme.spacing.unit}px`,
    rowGap: `${theme.spacing.unit}px`,
    gridRowGap: `${theme.spacing.unit}px`,
    gridAutoRows: '1fr'
  }
});

exports.default = (0, _styles.withStyles)(styles)(GridContent);
