// @pie-lib/scoring-config/lib/row.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _configUi = require('@pie-lib/config-ui');

var _IconButton = require('@material-ui/core/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _Delete = require('@material-ui/icons/Delete');

var _Delete2 = _interopRequireDefault(_Delete);

var _styles = require('@material-ui/core/styles');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Row = function Row(_ref) {
  var classes = _ref.classes,
      scorePercentage = _ref.scorePercentage,
      numberOfCorrect = _ref.numberOfCorrect,
      onRowChange = _ref.onRowChange,
      deletable = _ref.deletable,
      onDelete = _ref.onDelete,
      maxAnswers = _ref.maxAnswers;

  var onScoreChange = function onScoreChange(event, scorePercentage) {
    return onRowChange({ scorePercentage: scorePercentage, numberOfCorrect: numberOfCorrect });
  };
  var onNumberOfCorrectChange = function onNumberOfCorrectChange(event, numberOfCorrect) {
    return onRowChange({ scorePercentage: scorePercentage, numberOfCorrect: numberOfCorrect });
  };

  return _react2.default.createElement(
    'div',
    { className: classes.row },
    'Award',
    _react2.default.createElement(_configUi.NumberTextField, {
      className: classes.percent,
      min: 1,
      max: 99,
      value: scorePercentage,
      onChange: onScoreChange,
      suffix: '%'
    }),
    ' ',
    'for\xA0',
    _react2.default.createElement(_configUi.NumberTextField, {
      className: classes.field,
      min: 1,
      max: maxAnswers,
      value: numberOfCorrect,
      onChange: onNumberOfCorrectChange
    }),
    'correct answer',
    numberOfCorrect > 1 ? 's' : '',
    '.',
    deletable && _react2.default.createElement(
      _IconButton2.default,
      { className: classes.delete, onClick: onDelete },
      _react2.default.createElement(_Delete2.default, null)
    )
  );
};

var styles = function styles(theme) {
  return {
    row: {
      fontFamily: theme.typography.fontFamily,
      display: 'flex',
      alignItems: 'baseline'
    },
    percent: {
      width: '60px',
      paddingLeft: theme.spacing.unit
    },
    field: {
      width: '50px',
      paddingLeft: theme.spacing.unit
    },
    delete: {
      position: 'relative',
      top: '6px'
    }
  };
};

exports.default = (0, _styles.withStyles)(styles)(Row);


Row.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  numberOfCorrect: _propTypes2.default.number.isRequired,
  scorePercentage: _propTypes2.default.number.isRequired,
  onRowChange: _propTypes2.default.func.isRequired,
  deletable: _propTypes2.default.bool.isRequired,
  onDelete: _propTypes2.default.func.isRequired,
  maxAnswers: _propTypes2.default.number.isRequired
};
