// @pie-lib/config-ui/lib/tags-input/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TagsInput = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _uniq = require('lodash/uniq');

var _uniq2 = _interopRequireDefault(_uniq);

var _Chip = require('@material-ui/core/Chip');

var _Chip2 = _interopRequireDefault(_Chip);

var _muiBox = require('../mui-box');

var _muiBox2 = _interopRequireDefault(_muiBox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ENTER = 13;

var Tag = (0, _styles.withStyles)(function () {
  return {
    tag: {
      padding: '0px',
      margin: '1px'
    }
  };
})(function (_ref) {
  var classes = _ref.classes,
      label = _ref.label,
      onDelete = _ref.onDelete;
  return _react2.default.createElement(_Chip2.default, { className: classes.tag, label: label, onDelete: onDelete });
});

Tag.propTypes = {
  label: _propTypes2.default.string.isRequired,
  onDelete: _propTypes2.default.func.isRequired
};

var TagsInput = exports.TagsInput = function (_React$Component) {
  _inherits(TagsInput, _React$Component);

  function TagsInput(props) {
    _classCallCheck(this, TagsInput);

    var _this = _possibleConstructorReturn(this, (TagsInput.__proto__ || Object.getPrototypeOf(TagsInput)).call(this, props));

    _this.onFocus = function () {
      _this.setState({ focused: true });
    };

    _this.onBlur = function () {
      _this.setState({ focused: false });
    };

    _this.state = {
      value: '',
      focused: false
    };

    _this.onKeyDown = function (event) {
      if (event.keyCode === ENTER && _this.state.value !== '') {
        var tag = _this.state.value.trim();
        var newTags = (0, _uniq2.default)(_this.props.tags.concat([tag]));

        if (newTags.length !== _this.props.tags.length) {
          _this.props.onChange(newTags);
          _this.setState({ value: '' });
        }
      }
    };

    _this.onChange = function (event) {
      _this.setState({ value: event.target.value });
    };

    _this.deleteTag = function (tag) {
      var tags = _this.props.tags;


      var tagIndex = tags.indexOf(tag);
      if (tagIndex !== -1) {
        tags.splice(tagIndex, 1);
        _this.props.onChange(tags);
        _this.input.focus();
      }
    };
    return _this;
  }

  _createClass(TagsInput, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          classes = _props.classes,
          tags = _props.tags;

      return _react2.default.createElement(
        _muiBox2.default,
        { focused: this.state.focused },
        _react2.default.createElement(
          'div',
          { className: classes.tagsInput },
          (tags || []).map(function (t, index) {
            return _react2.default.createElement(Tag, { key: index, label: t, onDelete: function onDelete() {
                return _this2.deleteTag(t);
              } });
          }),
          _react2.default.createElement('input', {
            ref: function ref(r) {
              return _this2.input = r;
            },
            onKeyDown: this.onKeyDown,
            onChange: this.onChange,
            className: classes.input,
            value: this.state.value,
            onFocus: this.onFocus,
            onBlur: this.onBlur,
            type: 'text'
          })
        )
      );
    }
  }]);

  return TagsInput;
}(_react2.default.Component);

TagsInput.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  tags: _propTypes2.default.arrayOf(_propTypes2.default.string).isRequired,
  onChange: _propTypes2.default.func.isRequired
};


var styles = function styles(theme) {
  return {
    tagsInput: {
      border: 'solid 0px white',
      display: 'flex',
      flexWrap: 'wrap'
    },
    input: {
      padding: '2px',
      margin: '1px',
      minWidth: '30px',
      width: '100%',
      flex: '1',
      border: 'solid 0px white',
      height: '28px',
      fontSize: theme.typography.fontSize,
      fontFamily: theme.typography.fontFamily,
      outline: 'none',
      '&:focus': {
        outline: 'none'
      }
    }
  };
};

exports.default = (0, _styles.withStyles)(styles)(TagsInput);
