// @pie-element/select-text/configure/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _main = require('./main');

var _main2 = _interopRequireDefault(_main);

var _pieConfigureEvents = require('@pie-framework/pie-configure-events');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class SelectTextConfigure extends HTMLElement {
  constructor() {
    super();
  }

  set model(m) {
    this._model = m;
    this.render();
  }

  connectedCallback() {
    this.render();
  }

  modelChanged(m) {
    this._model = m;
    this.dispatchEvent(new _pieConfigureEvents.ModelUpdatedEvent(this._model), true);
    this.render();
  }

  render() {
    if (this._model) {
      const el = _react2.default.createElement(_main2.default, {
        model: this._model,
        onChange: this.modelChanged.bind(this)
      });

      _reactDom2.default.render(el, this);
    }
  }
}
exports.default = SelectTextConfigure;
