// @pie-element/categorize/configure/lib/scoring/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Scoring = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _weighting = require('./weighting');

var _weighting2 = _interopRequireDefault(_weighting);

var _partial = require('./partial');

var _partial2 = _interopRequireDefault(_partial);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const log = (0, _debug2.default)('@pie-element:categorize:configure:scoring');

class Scoring extends _react2.default.Component {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.changeWeighting = weighting => {
      log('[changeWeighting]', weighting);
      var _props = this.props;
      const scoring = _props.scoring,
            onChange = _props.onChange;

      scoring.weighting = weighting;
      onChange(scoring);
    }, this.changePartial = partial => {
      var _props2 = this.props;
      const scoring = _props2.scoring,
            onChange = _props2.onChange;

      scoring.partial = partial;
      onChange(scoring);
    }, _temp;
  }

  render() {
    var _props3 = this.props;
    const classes = _props3.classes,
          className = _props3.className,
          scoring = _props3.scoring,
          categories = _props3.categories,
          correctResponse = _props3.correctResponse;

    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(classes.scoring, className) },
      _react2.default.createElement(_weighting2.default, {
        weighting: scoring.weighting,
        onChange: this.changeWeighting,
        categories: categories
      }),
      _react2.default.createElement(_partial2.default, {
        className: classes.partial,
        partial: scoring.partial,
        correctResponse: correctResponse,
        categories: categories,
        onChange: this.changePartial
      })
    );
  }
}
exports.Scoring = Scoring;
Scoring.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string,
  scoring: _propTypes2.default.object,
  onChange: _propTypes2.default.func,
  categories: _propTypes2.default.array,
  correctResponse: _propTypes2.default.array
};
Scoring.defaultProps = {};
const styles = theme => ({
  scoring: {},
  partial: {
    marginTop: theme.spacing.unit * 2
  }
});

exports.default = (0, _styles.withStyles)(styles)(Scoring);
