// @pie-element/categorize/configure/lib/scoring/panel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Panel = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Typography = require('@material-ui/core/Typography');

var _Typography2 = _interopRequireDefault(_Typography);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _configUi = require('@pie-lib/config-ui');

var _buttons = require('../design/buttons');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Panel extends _react2.default.Component {
  render() {
    var _props = this.props;
    const classes = _props.classes,
          className = _props.className,
          children = _props.children,
          title = _props.title,
          enabled = _props.enabled,
          onToggleEnabled = _props.onToggleEnabled;

    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(classes.panel, className) },
      _react2.default.createElement(
        'div',
        { className: classes.header },
        _react2.default.createElement(
          _Typography2.default,
          { variant: 'title' },
          title
        ),
        _react2.default.createElement(_configUi.Checkbox, {
          label: 'Enabled',
          checked: enabled,
          onChange: onToggleEnabled
        })
      ),
      _react2.default.createElement(_buttons.Divider, null),
      children
    );
  }
}
exports.Panel = Panel;
Panel.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string,
  title: _propTypes2.default.string.isRequired,
  enabled: _propTypes2.default.bool.isRequired,
  onToggleEnabled: _propTypes2.default.func.isRequired,
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node]).isRequired
};
Panel.defaultProps = {};
const styles = () => ({
  panel: {},
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center'
  }
});
exports.default = (0, _styles.withStyles)(styles)(Panel);
