// @pie-ui/number-line/lib/number-line/transitions/fade.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _reactJss = require('react-jss');

var _reactJss2 = _interopRequireDefault(_reactJss);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactTransitionGroup = require('react-transition-group');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const duration = 200;
const fade = {
  appear: {
    opacity: 0
  },
  appearActive: {
    opacity: 1,
    transition: `opacity ${duration}ms ease-in`
  },
  enter: {
    opacity: 0
  },
  enterActive: {
    opacity: 1,
    transition: `opacity ${duration}ms ease-in`
  },
  exit: {
    opacity: 1
  },
  exitActive: {
    opacity: 0,
    transition: `opacity ${duration}ms ease-in`
  }
};

const FadeTransition = props => {
  const classes = props.classes;

  return _react2.default.createElement(_reactTransitionGroup.CSSTransition, _extends({}, props, {
    appear: true,
    classNames: {
      enter: classes.enter,
      enterActive: classes.enterActive,
      exit: classes.exit,
      exitActive: classes.exitActive,
      appear: classes.appear,
      appearActive: classes.appearActive
    },
    timeout: duration
  }));
};

FadeTransition.propTypes = {
  classes: _propTypes2.default.object.isRequired
};

const Fade = (0, _reactJss2.default)(fade)(FadeTransition);

exports.default = Fade;
