// @pie-element/categorize-configure/lib/main.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Main = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _configUi = require('@pie-lib/config-ui');

var _design = require('./design');

var _design2 = _interopRequireDefault(_design);

var _scoring = require('./scoring');

var _scoring2 = _interopRequireDefault(_scoring);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Main extends _react2.default.Component {
  constructor(...args) {
    var _temp;

    return _temp = super(...args), this.changeScoring = scoring => {
      var _props = this.props;
      const onChange = _props.onChange,
            model = _props.model;

      model.scoring = scoring;
      onChange(model);
    }, _temp;
  }

  render() {
    var _props2 = this.props;
    const classes = _props2.classes,
          className = _props2.className,
          model = _props2.model,
          onChange = _props2.onChange;

    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(classes.main, className) },
      _react2.default.createElement(
        _configUi.Tabs,
        null,
        _react2.default.createElement(_design2.default, { title: 'Design', model: model, onChange: onChange }),
        _react2.default.createElement(_scoring2.default, {
          title: 'Scoring',
          scoring: model.scoring,
          categories: model.categories,
          correctResponse: model.correctResponse,
          onChange: this.changeScoring
        })
      )
    );
  }
}
exports.Main = Main;
Main.propTypes = {
  classes: _propTypes2.default.object.isRequired,
  className: _propTypes2.default.string,
  model: _propTypes2.default.object.isRequired,
  onChange: _propTypes2.default.func.isRequired
};
Main.defaultProps = {};
const styles = theme => ({
  main: {}
});

const StyledMain = (0, _styles.withStyles)(styles)(Main);

class Stateful extends _react2.default.Component {

  constructor(props) {
    super(props);

    this.onChange = model => {
      this.setState({ model }, () => {
        this.props.onChange(this.state.model);
      });
    };

    this.state = {
      model: props.model
    };
  }

  componentWillReceiveProps(props) {
    this.setState({ model: props.model });
  }

  render() {
    return _react2.default.createElement(StyledMain, { model: this.state.model, onChange: this.onChange });
  }
}

Stateful.propTypes = {
  onChange: _propTypes2.default.func.isRequired,
  model: _propTypes2.default.object.isRequired
};
exports.default = Stateful;
