// @pie-lib/config-ui/lib/checkbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _FormControlLabel = require('@material-ui/core/FormControlLabel');

var _FormControlLabel2 = _interopRequireDefault(_FormControlLabel);

var _Checkbox = require('@material-ui/core/Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styles = require('@material-ui/core/styles');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _grey = require('@material-ui/core/colors/grey');

var _grey2 = _interopRequireDefault(_grey);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Checkbox = function Checkbox(_ref) {
  var mini = _ref.mini,
      checked = _ref.checked,
      onChange = _ref.onChange,
      value = _ref.value,
      label = _ref.label,
      classes = _ref.classes;
  return _react2.default.createElement(_FormControlLabel2.default, {
    className: (0, _classnames2.default)(classes.mini),
    classes: {
      label: (0, _classnames2.default)(classes.label, mini && classes.miniLabel)
    },
    control: _react2.default.createElement(_Checkbox2.default, {
      checked: checked,
      onChange: onChange,
      value: value,
      className: (0, _classnames2.default)(mini && classes.miniCheckbox)
    }),
    label: label
  });
};

Checkbox.propTypes = {
  mini: _propTypes2.default.bool,
  classes: _propTypes2.default.object.isRequired,
  checked: _propTypes2.default.bool.isRequired,
  onChange: _propTypes2.default.func.isRequired,
  value: _propTypes2.default.string,
  label: _propTypes2.default.string.isRequired
};

Checkbox.defaultProps = {
  value: '',
  mini: false
};

exports.default = (0, _styles.withStyles)(function (theme) {
  return {
    label: {
      fontSize: '13px',
      transform: 'translate(-4%, 2%)',
      color: 'rgba(0,0,0,1.0)'
    },
    miniCheckbox: {
      margin: 0,
      padding: 0,
      width: theme.spacing.unit * 3,
      height: theme.spacing.unit * 3
    },
    miniLabel: {
      marginLeft: theme.spacing.unit,
      color: _grey2.default[700],
      fontSize: '11px'
    },
    mini: {
      margin: 0,
      marginLeft: 0,
      padding: 0
    }
  };
})(Checkbox);
