// @pie-lib/config-ui/lib/two-choice.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NChoice = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _inputContainer = require('./input-container');

var _inputContainer2 = _interopRequireDefault(_inputContainer);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _radioWithLabel = require('./radio-with-label');

var _radioWithLabel2 = _interopRequireDefault(_radioWithLabel);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styles = require('@material-ui/core/styles');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = function styles(theme) {
  return {
    group: {
      display: 'flex',
      paddingLeft: 0,
      marginTop: theme.spacing.unit
    },
    vertical: {
      flexDirection: 'column'
    }
  };
};

var RawNChoice = function (_React$Component) {
  _inherits(RawNChoice, _React$Component);

  function RawNChoice() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, RawNChoice);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = RawNChoice.__proto__ || Object.getPrototypeOf(RawNChoice)).call.apply(_ref, [this].concat(args))), _this), _this.handleChange = function (event) {
      _this.props.onChange(event.currentTarget.value);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(RawNChoice, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          header = _props.header,
          className = _props.className,
          classes = _props.classes,
          opts = _props.opts,
          value = _props.value,
          direction = _props.direction;


      return _react2.default.createElement(
        _inputContainer2.default,
        { label: header, className: className },
        _react2.default.createElement(
          'div',
          { className: (0, _classnames2.default)(classes.group, classes[direction]) },
          opts.map(function (o, index) {
            return _react2.default.createElement(_radioWithLabel2.default, {
              value: o.value,
              key: index,
              checked: o.value === value,
              onChange: _this2.handleChange,
              label: o.label
            });
          })
        )
      );
    }
  }]);

  return RawNChoice;
}(_react2.default.Component);

RawNChoice.propTypes = {
  header: _propTypes2.default.string.isRequired,
  className: _propTypes2.default.string,
  opts: _propTypes2.default.array.isRequired,
  value: _propTypes2.default.string,
  onChange: _propTypes2.default.func.isRequired,
  direction: _propTypes2.default.oneOf(['horizontal', 'vertical']),
  classes: _propTypes2.default.object.isRequired
};
var NChoice = exports.NChoice = (0, _styles.withStyles)(styles)(RawNChoice);

var TwoChoice = function (_React$Component2) {
  _inherits(TwoChoice, _React$Component2);

  function TwoChoice() {
    _classCallCheck(this, TwoChoice);

    return _possibleConstructorReturn(this, (TwoChoice.__proto__ || Object.getPrototypeOf(TwoChoice)).apply(this, arguments));
  }

  _createClass(TwoChoice, [{
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          one = _props2.one,
          two = _props2.two,
          header = _props2.header,
          className = _props2.className,
          value = _props2.value,
          onChange = _props2.onChange;

      var opts = [one, two];
      return _react2.default.createElement(NChoice, {
        header: header,
        className: className,
        opts: opts,
        value: value,
        onChange: onChange
      });
    }
  }]);

  return TwoChoice;
}(_react2.default.Component);

TwoChoice.propTypes = {
  header: _propTypes2.default.string.isRequired,
  value: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func.isRequired,
  one: _propTypes2.default.shape({
    label: _propTypes2.default.string.isRequired,
    value: _propTypes2.default.string.isRequired
  }),
  two: _propTypes2.default.shape({
    label: _propTypes2.default.string.isRequired,
    value: _propTypes2.default.string.isRequired
  }),
  className: _propTypes2.default.string
};
exports.default = (0, _styles.withStyles)(styles)(TwoChoice);
