// @pie-lib/config-ui/lib/with-stateful-model.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var withStatefulModel = function withStatefulModel(Component) {
  var Stateful = function (_React$Component) {
    _inherits(Stateful, _React$Component);

    function Stateful(props) {
      _classCallCheck(this, Stateful);

      var _this = _possibleConstructorReturn(this, (Stateful.__proto__ || Object.getPrototypeOf(Stateful)).call(this, props));

      _this.onChange = function (model) {
        _this.setState({ model: model }, function () {
          _this.props.onChange(_this.state.model);
        });
      };

      _this.state = {
        model: props.model
      };
      return _this;
    }

    _createClass(Stateful, [{
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(props) {
        this.setState({ model: props.model });
      }
    }, {
      key: 'render',
      value: function render() {
        return _react2.default.createElement(Component, { model: this.state.model, onChange: this.onChange });
      }
    }]);

    return Stateful;
  }(_react2.default.Component);

  Stateful.propTypes = {
    model: _propTypes2.default.object.isRequired,
    onChange: _propTypes2.default.func.isRequired
  };


  return Stateful;
};

exports.default = withStatefulModel;
