// @vx/text/build/text/TextBackground.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TextBackground;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function TextBackground(_ref) {
  var _ref$x = _ref.x,
      x = _ref$x === undefined ? 0 : _ref$x,
      _ref$y = _ref.y,
      y = _ref$y === undefined ? 0 : _ref$y,
      _ref$dx = _ref.dx,
      dx = _ref$dx === undefined ? 0 : _ref$dx,
      _ref$dy = _ref.dy,
      dy = _ref$dy === undefined ? 0 : _ref$dy,
      children = _ref.children,
      _ref$fontSize = _ref.fontSize,
      fontSize = _ref$fontSize === undefined ? 12 : _ref$fontSize,
      _ref$fontFamily = _ref.fontFamily,
      fontFamily = _ref$fontFamily === undefined ? 'Arial' : _ref$fontFamily,
      _ref$textAnchor = _ref.textAnchor,
      textAnchor = _ref$textAnchor === undefined ? 'start' : _ref$textAnchor,
      _ref$fill = _ref.fill,
      fill = _ref$fill === undefined ? 'white' : _ref$fill,
      _ref$stroke = _ref.stroke,
      stroke = _ref$stroke === undefined ? 'none' : _ref$stroke,
      _ref$strokeWidth = _ref.strokeWidth,
      strokeWidth = _ref$strokeWidth === undefined ? 0 : _ref$strokeWidth,
      strokeDasharray = _ref.strokeDasharray,
      _ref$backgroundFill = _ref.backgroundFill,
      backgroundFill = _ref$backgroundFill === undefined ? 'white' : _ref$backgroundFill,
      _ref$backgroundStroke = _ref.backgroundStroke,
      backgroundStroke = _ref$backgroundStroke === undefined ? 'magenta' : _ref$backgroundStroke,
      _ref$backgroundStroke2 = _ref.backgroundStrokeWidth,
      backgroundStrokeWidth = _ref$backgroundStroke2 === undefined ? 3 : _ref$backgroundStroke2,
      backgroundStrokeDasharray = _ref.backgroundStrokeDasharray,
      className = _ref.className;

  return _react2.default.createElement(
    'g',
    null,
    _react2.default.createElement('rect', {
      className: (0, _classnames2.default)('vx-text-background__background', className),
      width: '100%',
      height: '100%',
      fill: backgroundFill,
      stroke: backgroundStroke,
      strokeWidth: backgroundStrokeWidth,
      strokeDasharray: backgroundStrokeDasharray
    }),
    _react2.default.createElement(
      'text',
      {
        className: (0, _classnames2.default)('vx-text-background__text', className),
        x: x,
        y: y,
        dx: dx,
        dy: dy,
        fontSize: fontSize,
        fontFamily: fontFamily,
        fill: fill,
        stroke: stroke,
        strokeWidth: strokeWidth,
        strokeDasharray: strokeDasharray,
        textAnchor: textAnchor
      },
      children
    )
  );
}
