// @vx/text/build/text/TextOutline.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TextOutline;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function TextOutline(_ref) {
  var _ref$x = _ref.x,
      x = _ref$x === undefined ? 0 : _ref$x,
      _ref$y = _ref.y,
      y = _ref$y === undefined ? 0 : _ref$y,
      _ref$dx = _ref.dx,
      dx = _ref$dx === undefined ? 0 : _ref$dx,
      _ref$dy = _ref.dy,
      dy = _ref$dy === undefined ? 0 : _ref$dy,
      children = _ref.children,
      _ref$fontSize = _ref.fontSize,
      fontSize = _ref$fontSize === undefined ? 12 : _ref$fontSize,
      _ref$fontFamily = _ref.fontFamily,
      fontFamily = _ref$fontFamily === undefined ? 'Arial' : _ref$fontFamily,
      _ref$fill = _ref.fill,
      fill = _ref$fill === undefined ? 'white' : _ref$fill,
      _ref$stroke = _ref.stroke,
      stroke = _ref$stroke === undefined ? 'none' : _ref$stroke,
      _ref$strokeWidth = _ref.strokeWidth,
      strokeWidth = _ref$strokeWidth === undefined ? 0 : _ref$strokeWidth,
      _ref$outlineFill = _ref.outlineFill,
      outlineFill = _ref$outlineFill === undefined ? 'white' : _ref$outlineFill,
      _ref$outlineStroke = _ref.outlineStroke,
      outlineStroke = _ref$outlineStroke === undefined ? 'magenta' : _ref$outlineStroke,
      _ref$outlineStrokeWid = _ref.outlineStrokeWidth,
      outlineStrokeWidth = _ref$outlineStrokeWid === undefined ? 3 : _ref$outlineStrokeWid,
      _ref$textAnchor = _ref.textAnchor,
      textAnchor = _ref$textAnchor === undefined ? 'start' : _ref$textAnchor,
      className = _ref.className;

  return _react2.default.createElement(
    'g',
    null,
    _react2.default.createElement(
      'text',
      {
        className: (0, _classnames2.default)('vx-text-outline__outline', className),
        x: x,
        y: y,
        dx: dx,
        dy: dy,
        fill: outlineFill,
        fontSize: fontSize,
        fontFamily: fontFamily,
        stroke: outlineStroke,
        strokeWidth: outlineStrokeWidth,
        textAnchor: textAnchor
      },
      children
    ),
    _react2.default.createElement(
      'text',
      {
        className: (0, _classnames2.default)('vx-text-outline__text', className),
        x: x,
        y: y,
        dx: dx,
        dy: dy,
        fontSize: fontSize,
        fontFamily: fontFamily,
        fill: fill,
        stroke: stroke,
        strokeWidth: strokeWidth,
        textAnchor: textAnchor
      },
      children
    )
  );
}
