// @pie-lib/render-ui/lib/response-indicators.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NothingSubmitted = exports.PartiallyCorrect = exports.Incorrect = exports.Correct = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _icons = require('@pie-lib/icons');

var icons = _interopRequireWildcard(_icons);

var _Popover = require('@material-ui/core/Popover');

var _Popover2 = _interopRequireDefault(_Popover);

var _styles = require('@material-ui/core/styles');

var _feedback = require('./feedback');

var _feedback2 = _interopRequireDefault(_feedback);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var log = (0, _debug2.default)('pie-libs:render-ui:response-indicators');

var styles = function styles() {
  return {
    responseIndicator: {
      cursor: 'pointer'
    },
    paper: {
      padding: '0',
      borderRadius: '4px'
    },
    popover: {
      cursor: 'pointer'
    },
    popperClose: {
      cursor: 'pointer'
    }
  };
};

var BuildIndicator = function BuildIndicator(Icon, correctness) {
  var RawIndicator = function (_React$Component) {
    _inherits(RawIndicator, _React$Component);

    function RawIndicator(props) {
      _classCallCheck(this, RawIndicator);

      var _this = _possibleConstructorReturn(this, (RawIndicator.__proto__ || Object.getPrototypeOf(RawIndicator)).call(this, props));

      _this.handlePopoverOpen = function (event) {
        log('[handlePopoverOpen]', event.target);
        _this.setState({ anchorEl: event.target });
      };

      _this.handlePopoverClose = function () {
        _this.setState({ anchorEl: null });
      };

      _this.state = {};
      return _this;
    }

    _createClass(RawIndicator, [{
      key: 'render',
      value: function render() {
        var _this2 = this;

        var _props = this.props,
            feedback = _props.feedback,
            classes = _props.classes;
        var anchorEl = this.state.anchorEl;

        return _react2.default.createElement(
          'div',
          { className: feedback && classes.responseIndicator },
          _react2.default.createElement(
            'span',
            { ref: function ref(r) {
                return _this2.icon = r;
              }, onClick: this.handlePopoverOpen },
            _react2.default.createElement(Icon, null)
          ),
          feedback && _react2.default.createElement(
            _Popover2.default,
            {
              className: classes.popover,
              classes: {
                paper: classes.paper
              },
              open: !!anchorEl,
              anchorEl: anchorEl,
              anchorOrigin: {
                vertical: 'bottom',
                horizontal: 'left'
              },
              transformOrigin: {
                vertical: 'top',
                horizontal: 'left'
              },
              onClose: this.handlePopoverClose
            },
            _react2.default.createElement(_feedback2.default, { feedback: feedback, correctness: correctness })
          )
        );
      }
    }]);

    return RawIndicator;
  }(_react2.default.Component);

  RawIndicator.propTypes = {
    feedback: _propTypes2.default.string,
    classes: _propTypes2.default.object.isRequired
  };

  return (0, _styles.withStyles)(styles)(RawIndicator);
};

var Correct = exports.Correct = BuildIndicator(icons.Correct, 'correct');
var Incorrect = exports.Incorrect = BuildIndicator(icons.Incorrect, 'incorrect');
var PartiallyCorrect = exports.PartiallyCorrect = BuildIndicator(icons.PartiallyCorrect, 'partially-correct');
var NothingSubmitted = exports.NothingSubmitted = BuildIndicator(icons.NothingSubmitted, 'nothing-submitted');
